connector/src/main/java/com/google/cloud/spark/bigquery/AvroSchemaConverter.java

```diff
@@ -0,0 +1,287 @@
+    }
+    if (sparkType instanceof BinaryType && avroType.getType() == Schema.Type.FIXED) {
+      int size = avroType.getFixedSize();
+      return (getter, ordinal) -> {
+        byte[] data = getter.getBinary(ordinal);
+        if (data.length != size) {
+          throw new IllegalArgumentException(
+              String.format(
+                  "Cannot write %s bytes of binary data into FIXED Type with size of %s bytes",
+                  data.length, size));
+        }
+        return new GenericData.Fixed(avroType, data);
+      };
+    }
+    if (sparkType instanceof BinaryType && avroType.getType() == Schema.Type.BYTES) {
+      return (getter, ordinal) -> ByteBuffer.wrap(getter.getBinary(ordinal));
+    }
+
+    if (sparkType instanceof DateType && avroType.getType() == Schema.Type.INT) {
+      return (getter, ordinal) -> getter.getInt(ordinal);
+    }
+
+    if (sparkType instanceof TimestampType && avroType.getType() == Schema.Type.LONG) {
+      return (getter, ordinal) -> getter.getLong(ordinal);
+    }
+
+    if (sparkType instanceof ArrayType && avroType.getType() == Schema.Type.ARRAY) {
+      DataType et = ((ArrayType) sparkType).elementType();
+      boolean containsNull = ((ArrayType) sparkType).containsNull();
+
+      Converter elementConverter =
+          createConverterFor(et, resolveNullableType(avroType.getElementType(), containsNull));
+      return (getter, ordinal) -> {
+        ArrayData arrayData = getter.getArray(ordinal);
+        int len = arrayData.numElements();
+        Object[] result = new Object[len];
+        for (int i = 0; i < len; i++) {
+          if (containsNull && arrayData.isNullAt(i)) {
+            result[i] = null;
+          } else {
+            result[i] = elementConverter.convert(arrayData, i);
+          }
+        }
+        // avro writer is expecting a Java Collection, so we convert it into
+        // `ArrayList` backed by the specified array without data copying.
+        return java.util.Arrays.asList(result);
+      };
+    }
+    if (sparkType instanceof StructType && avroType.getType() == Schema.Type.RECORD) {
+      StructType sparkStruct = (StructType) sparkType;
+
+      StructConverter structConverter = new StructConverter(sparkStruct, avroType);
+      int numFields = sparkStruct.length();
+      return (getter, ordinal) -> structConverter.convert(getter.getStruct(ordinal, numFields));
+    }
+    if (sparkType instanceof UserDefinedType) {
+      UserDefinedType userDefinedType = (UserDefinedType) sparkType;
+      return createConverterFor(userDefinedType.sqlType(), avroType);
+    }
+    throw new IllegalArgumentException(
+        String.format("Cannot convert Catalyst type %s to Avro type %s", sparkType, avroType));
+  }
+
+  @FunctionalInterface
+  interface Converter {
+    // void convert(InternalRow fromRow, GenericData.Record toRecord, int index);
+    Object convert(SpecializedGetters getters, int ordinal);
```

Comment:
Should we have a generic interface for type safety?

Reply:
No - the only place where it is used is in `GenericData.Record.put()` which accepts `Object`. There is no meaning for generics here.